homeassistant/components/tuya.py

```diff
@@ -0,0 +1,166 @@
+            hass, ha_type, DOMAIN,
+            {'dev_ids': device_type_list.get(ha_type)}, config)
+
+    def poll_devices_update(event_time):
+        """Check if accesstoken is expired and pull device list from server."""
+        _LOGGER.info("Pull Devices From Tuya")
+        devices = tuya.poll_devices_update()
+        if devices is None:
+            return
+        # Add new discover device.
+        device_list = tuya.get_all_devices()
+        device_type_list = {}
+        for device in device_list:
+            dev_type = device.device_type()
+            if (dev_type in TUYA_TYPE_TO_HA and
+                    device.object_id() not in hass.data[DOMAIN]['entities']):
+                ha_type = TUYA_TYPE_TO_HA.get(dev_type)
+                if ha_type not in device_type_list:
+                    device_type_list[ha_type] = []
+                device_type_list[ha_type].append(device.object_id())
+        for ha_type in device_type_list:
+            discovery.load_platform(
+                hass, ha_type, DOMAIN,
+                {'dev_ids': device_type_list.get(ha_type)}, config)
+        # Delete not exist device.
+        newlist_ids = []
+        for device in devices:
+            newlist_ids.append(device.get('id'))
+        for dev_id in list(hass.data[DOMAIN]['entities']):
+            if dev_id not in newlist_ids:
+                dispatcher_send(hass, SIGNAL_DELETE_ENTITY, dev_id)
+                hass.data[DOMAIN]['entities'].pop(dev_id)
+
+    track_time_interval(hass, poll_devices_update, timedelta(minutes=5))
+
+    hass.services.register(DOMAIN, SERVICE_PULL_DEVICES, poll_devices_update)
+
+    def force_update(call):
+        """Force all devices to pull data."""
+        _LOGGER.info("Refreshing Device Data From Tuya")
+        dispatcher_send(hass, SIGNAL_UPDATE_ENTITY)
+
+    hass.services.register(DOMAIN, SERVICE_FORCE_UPDATE, force_update)
+
+    return True
+
+
+class TuyaDevice(Entity):
+    """Tuya base device."""
+
+    def __init__(self, tuya):
+        """Init Tuya devices."""
+        self.tuya = tuya
+
+    @property
+    def object_id(self):
+        """Return Tuya device id."""
+        return self.tuya.object_id()
+
+    @property
+    def name(self):
+        """Return Tuya device name."""
+        return self.tuya.name()
+
+    @property
+    def is_on(self):
+        """Return true if device is on."""
+        return self.tuya.state()
+
+    @property
+    def icon(self):
+        """Return the entity picture to use in the frontend, if any."""
+        return self.tuya.iconurl()
+
+    @property
+    def available(self):
+        """Return if the device is available."""
+        return self.tuya.available()
+
+    def update(self):
+        """Refresh Tuya device data."""
+        self.tuya.update()
+
+    @callback
+    def _delete_callback(self, dev_id):
+        """Remove this entity."""
+        if dev_id == self.object_id:
+            self.hass.add_job(self.async_remove())
```

Comment:
We can use `self.hass.async_add_job` here since we're in a callback that will run in the event loop.